File: whatsapp/app/scripts/generate-icons.js
import fs from 'fs';
import path from 'path';
import { Jimp } from 'jimp';

async function generatePngIcons() {
  const sizes = [16, 32, 48, 72, 96, 128, 144, 152, 180, 192, 384, 512];
  const uiDir = path.resolve('src/routes/ui');

  for (const size of sizes) {
    const image = new Jimp({ width: size, height: size, color: 0x00a884ff });

    const cx = Math.floor(size / 2);
    const cy = Math.floor(size / 2);
    const r = Math.floor(size * 0.35);

    for (let x = 0; x < size; x++) {
      for (let y = 0; y < size; y++) {
        const dist = Math.sqrt((x - cx) ** 2 + (y - cy) ** 2);
        if (dist <= r && dist >= r - Math.max(1, Math.floor(size * 0.08))) {
          image.setPixelColor(0xffffffff, x, y);
        }
        const cornerDist = Math.max(
          Math.abs(x - cx) - (size / 2 - size * 0.2),
          Math.abs(y - cy) - (size / 2 - size * 0.2),
          0
        );
        if (cornerDist > size * 0.2) {
          image.setPixelColor(0x00000000, x, y);
        }
      }
    }

    const outPath = path.join(uiDir, `icon-${size}.png`);
    const buffer = await image.getBuffer('image/png');
    fs.writeFileSync(outPath, buffer);
  }
  console.log('✅ Generated PNG icon set successfully');
}

generatePngIcons().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: whatsapp/app/scripts/generate-icons.py
"""Generate all PWA and UI icons from the official WhatsApp Add-on logo.

Usage:
    python scripts/generate-icons.py
"""

import os
from pathlib import Path

from PIL import Image


def generate_icons():
    root_dir = Path(__file__).resolve().parent.parent
    logo_path = root_dir.parent / "logo.png"
    ui_dir = root_dir / "src" / "routes" / "ui"

    if not logo_path.exists():
        print(f"Error: Logo file not found at {logo_path}")
        return

    img = Image.open(logo_path).convert("RGBA")
    print(f"Loaded official logo: {img.size} from {logo_path}")

    sizes = [16, 32, 48, 72, 96, 128, 144, 152, 180, 192, 384, 512]

    for s in sizes:
        target_path = ui_dir / f"icon-{s}.png"
        resized = img.resize((s, s), Image.Resampling.LANCZOS)
        resized.save(target_path, format="PNG")
        print(f"Generated {target_path.name} ({s}x{s})")

    # Multi-resolution favicon.ico
    ico_path = ui_dir / "favicon.ico"
    img.resize((48, 48), Image.Resampling.LANCZOS).save(ico_path, format="ICO", sizes=[(16, 16), (32, 32), (48, 48)])
    print("Generated favicon.ico (16, 32, 48)")


if __name__ == "__main__":
    generate_icons()
